File: rxbinding/src/main/java/com/jakewharton/rxbinding2/view/ViewTreeObserverGlobalLayoutObservable.java
package com.jakewharton.rxbinding2.view;

import android.os.Build;
import android.view.View;
import android.view.ViewTreeObserver.OnGlobalLayoutListener;
import com.jakewharton.rxbinding2.internal.Notification;
import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.android.MainThreadDisposable;

import static com.jakewharton.rxbinding2.internal.Preconditions.checkMainThread;

final class ViewTreeObserverGlobalLayoutObservable extends Observable<Object> {
  private final View view;

  ViewTreeObserverGlobalLayoutObservable(View view) {
    this.view = view;
  }

  @Override protected void subscribeActual(Observer<? super Object> observer) {
    if (!checkMainThread(observer)) {
      return;
    }
    Listener listener = new Listener(view, observer);
    observer.onSubscribe(listener);
    view.getViewTreeObserver()
        .addOnGlobalLayoutListener(listener);
  }

  static final class Listener extends MainThreadDisposable implements OnGlobalLayoutListener {
    private final View view;
    private final Observer<? super Object> observer;

    Listener(View view, Observer<? super Object> observer) {
      this.view = view;
      this.observer = observer;
    }

    @Override public void onGlobalLayout() {
      if (!isDisposed()) {
        observer.onNext(Notification.INSTANCE);
      }
    }

    @Override protected void onDispose() {
      if (Build.VERSION.SDK_INT > Build.VERSION_CODES.ICE_CREAM_SANDWICH_MR1) {
        view.getViewTreeObserver().removeOnGlobalLayoutListener(this);
      } else {
        view.getViewTreeObserver().removeGlobalOnLayoutListener(this);
      }
    }
  }
}

File: rxbinding/src/main/java/com/jakewharton/rxbinding4/view/ViewTreeObserverGlobalLayoutObservable.kt
@file:JvmName("RxView")
@file:JvmMultifileClass

package com.jakewharton.rxbinding4.view

import android.view.View
import android.view.ViewTreeObserver
import android.view.ViewTreeObserver.OnGlobalLayoutListener
import androidx.annotation.CheckResult
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.core.Observer
import io.reactivex.rxjava3.android.MainThreadDisposable

import com.jakewharton.rxbinding4.internal.checkMainThread

/**
 * Create an observable which emits on `view` globalLayout events. The emitted value is
 * unspecified and should only be used as notification.
 *
 * *Warning:* The created observable keeps a strong reference to `view`. Unsubscribe
 * to free this reference.
 *
 * *Warning:* The created observable uses [ ][ViewTreeObserver.addOnGlobalLayoutListener] to observe global layouts. Multiple observables
 * can be used for a view at a time.
 */
@CheckResult
fun View.globalLayouts(): Observable<Unit> {
  return ViewTreeObserverGlobalLayoutObservable(this)
}

private class ViewTreeObserverGlobalLayoutObservable(
  private val view: View
) : Observable<Unit>() {

  override fun subscribeActual(observer: Observer<in Unit>) {
    if (!checkMainThread(observer)) {
      return
    }
    val listener = Listener(view, observer)
    observer.onSubscribe(listener)
    view.viewTreeObserver
        .addOnGlobalLayoutListener(listener)
  }

  private class Listener(
    private val view: View,
    private val observer: Observer<in Unit>
  ) : MainThreadDisposable(), OnGlobalLayoutListener {

    override fun onGlobalLayout() {
      if (!isDisposed) {
        observer.onNext(Unit)
      }
    }

    override fun onDispose() {
      @Suppress("DEPRECATION") // Correct when minSdk 16.
      view.viewTreeObserver.removeGlobalOnLayoutListener(this)
    }
  }
}
